File: swex/tutorials/geometrical-primitives/vb-net/PrimitivesStandardFeaturesPMPage/PrimitivesStandardFeaturesPMPage/ModelDocEx.vb
'**********************
'SwEx - development tools for SOLIDWORKS
'Copyright(C) 2019 www.codestack.net
'License: https://github.com/codestackdev/swex-common/blob/master/LICENSE
'Product URL: https://www.codestack.net/labs/solidworks/swex
'**********************

Imports SolidWorks.Interop.swconst
Imports System
Imports System.Runtime.CompilerServices
Imports SolidWorks.Interop.sldworks

Public Module ModelDocEx

    <Extension()>
    Sub CreateBox(ByVal model As IModelDoc2, ByVal reference As IEntity, ByVal width As Double, ByVal length As Double, ByVal height As Double)
        CreateExtrudedPrimitive(model, reference, Function(skMgr) skMgr.CreateCenterRectangle(0, 0, 0, width / 2, height / 2, 0) IsNot Nothing, height)
    End Sub

    <Extension()>
    Sub CreateCylinder(ByVal model As IModelDoc2, ByVal reference As IEntity, ByVal diam As Double, ByVal height As Double)
        CreateExtrudedPrimitive(model, reference, Function(skMgr) skMgr.CreateCircleByRadius(0, 0, 0, diam / 2) IsNot Nothing, height)
    End Sub

    Private Sub CreateExtrudedPrimitive(ByVal model As IModelDoc2, ByVal reference As IEntity, ByVal creator As Func(Of ISketchManager, Boolean), ByVal height As Double)
        Try

            If Not reference.SelectByMark(False, 0) Then
                Throw New Exception("Failed to select reference")
            End If

            model.IActiveView.EnableGraphicsUpdate = False
            model.FeatureManager.EnableFeatureTree = False
            Dim sketch = CreateSketch(model, creator)
            ExtrudeSketch(model, sketch, height)
        Finally
            model.IActiveView.EnableGraphicsUpdate = True
            model.FeatureManager.EnableFeatureTree = True
        End Try
    End Sub

    Private Function CreateSketch(ByVal model As IModelDoc2, ByVal creator As Func(Of ISketchManager, Boolean)) As Sketch

        If model Is Nothing Then
            Throw New NullReferenceException("Model is null")
        End If

        Dim skMgr = model.SketchManager
        skMgr.InsertSketch(True)
        skMgr.AddToDB = True
        Dim sketch = skMgr.ActiveSketch

        If sketch Is Nothing Then
            Throw New NullReferenceException("Sketch is null")
        End If

        If Not creator(skMgr) Then
            Throw New NullReferenceException("Failed to create rectangle")
        End If

        skMgr.AddToDB = False
        skMgr.InsertSketch(True)

        Return sketch

    End Function

    Private Sub ExtrudeSketch(ByVal model As IModelDoc2, ByVal sketch As Sketch, ByVal height As Double)

        If (TryCast(sketch, IFeature)).Select2(False, 0) Then
            Dim feat = model.FeatureManager.FeatureExtrusion3(True, False, False, CInt(swEndConditions_e.swEndCondBlind), CInt(CInt(swEndConditions_e.swEndCondBlind)), height, 0, False, False, False, False, 0, 0, False, False, False, False, False, False, False, CInt(swStartConditions_e.swStartSketchPlane), 0, False)

            If feat Is Nothing Then
                Throw New NullReferenceException("Failed to create extrusion feature")
            End If
        Else
            Throw New Exception("Failed to select sketch")
        End If

    End Sub

End Module

File: swex/tutorials/geometrical-primitives/csharp/PrimitivesStandardFeaturesPMPage/PrimitivesStandardFeaturesPMPage/ModelDocEx.cs
//**********************
//SwEx - development tools for SOLIDWORKS
//Copyright(C) 2019 www.codestack.net
//License: https://github.com/codestackdev/swex-common/blob/master/LICENSE
//Product URL: https://www.codestack.net/labs/solidworks/swex
//**********************

using SolidWorks.Interop.swconst;
using System;
using System.Runtime.InteropServices;

namespace SolidWorks.Interop.sldworks
{
    public static class ModelDocEx
    {
        public static void CreateBox(this IModelDoc2 model, IEntity reference, double width, double length, double height)
        {
            CreateExtrudedPrimitive(model, reference, skMgr => skMgr.CreateCenterRectangle(0, 0, 0, width / 2, height / 2, 0) != null, height);
        }

        public static void CreateCylinder(this IModelDoc2 model, IEntity reference, double diam, double height)
        {
            CreateExtrudedPrimitive(model, reference, skMgr => skMgr.CreateCircleByRadius(0, 0, 0, diam / 2) != null, height);
        }

        private static void CreateExtrudedPrimitive(IModelDoc2 model, IEntity reference, Func<ISketchManager, bool> creator, double height)
        {
            try
            {
                if (!reference.SelectByMark(false, 0))
                {
                    throw new Exception("Failed to select reference");
                }

                model.IActiveView.EnableGraphicsUpdate = false;
                model.FeatureManager.EnableFeatureTree = false;

                var sketch = CreateSketch(model, creator);

                ExtrudeSketch(model, sketch, height);
            }
            finally
            {
                model.IActiveView.EnableGraphicsUpdate = true;
                model.FeatureManager.EnableFeatureTree = true;
            }
        }

        private static Sketch CreateSketch(IModelDoc2 model, Func<ISketchManager, bool> creator)
        {
            if (model == null)
            {
                throw new NullReferenceException("Model is null");
            }

            var skMgr = model.SketchManager;
            
            skMgr.InsertSketch(true);
            skMgr.AddToDB = true;

            var sketch = skMgr.ActiveSketch;

            if (sketch == null)
            {
                throw new NullReferenceException("Sketch is null");
            }

            if (!creator.Invoke(skMgr))
            {
                throw new NullReferenceException("Failed to create rectangle");
            }

            skMgr.AddToDB = false;
            skMgr.InsertSketch(true);

            return sketch;
        }

        private static void ExtrudeSketch(IModelDoc2 model, Sketch sketch, double height)
        {
            if ((sketch as IFeature).Select2(false, 0))
            {
                var feat = model.FeatureManager.FeatureExtrusion3(true, false, false,
                    (int)swEndConditions_e.swEndCondBlind, (int)(int)swEndConditions_e.swEndCondBlind,
                    height, 0, false, false, false, false, 0, 0, false, false, false, false, false, false, false,
                    (int)swStartConditions_e.swStartSketchPlane, 0, false);

                if (feat == null)
                {
                    throw new NullReferenceException("Failed to create extrusion feature");
                }
            }
            else
            {
                throw new Exception("Failed to select sketch");
            }
        }
    }
}
